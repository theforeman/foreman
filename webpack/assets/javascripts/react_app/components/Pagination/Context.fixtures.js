const contextFixtures = {
  metadata: {
    UISettings: { perPage: 5, perPageOptions: [5, 10, 15, 25, 50] },
  },
};

export default contextFixtures;
