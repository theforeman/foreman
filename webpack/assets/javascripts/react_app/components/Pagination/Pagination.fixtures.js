export const paginationMock = {
  data: {
    viewType: 'table',
    perPage: 20,
    perPageOptions: [5, 10, 15, 20],
    itemCount: 21,
  },
};
