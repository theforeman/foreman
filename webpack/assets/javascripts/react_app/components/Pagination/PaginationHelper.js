import URI from 'urijs';

export const getURI = () => new URI(window.location.href);

export const getURIpage = () => Number(getURI().query(true).page);
export const getURIperPage = () => Number(getURI().query(true).per_page);
export const getURIsearch = () => getURI().query(true).search;

export const changeQuery = (uri, newQuery, navigateTo) => {
  uri.setQuery(newQuery);
  if (navigateTo) {
    navigateTo(uri.toString());
    return uri.toString();
  }
  return window.Turbolinks.visit(uri.toString());
};
