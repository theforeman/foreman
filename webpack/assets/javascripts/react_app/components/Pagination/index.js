import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { useHistory } from 'react-router-dom';
import URI from 'urijs';
import classNames from 'classnames';
import {
  Pagination as PF4Pagination,
  PaginationVariant,
} from '@patternfly/react-core';
import { translate as __ } from '../../common/I18n';
import { useForemanSettings } from '../../Root/Context/ForemanContext';
import {
  getURIpage,
  getURIperPage,
  changeQuery,
} from '../../common/urlHelpers';
import './index.scss';

const Pagination = ({
  itemCount,
  onSetPage,
  onPerPageSelect,
  onChange,
  className,
  page: propsPage,
  perPage: propsPerPage,
  noSidePadding,
  variant,
  updateParamsByUrl,
  ...props
}) => {
  const { perPage: settingsPerPage = 20 } = useForemanSettings() || {};
  const [page, setPage] = useState(propsPage);
  const [perPage, setPerPage] = useState(propsPerPage || settingsPerPage);
  const history = useHistory();
  const { location: { search } = {} } = history || {};

  useEffect(() => {
    let nextPage = propsPage;
    let nextPerPage = propsPerPage;
    if (updateParamsByUrl) {
      if (search !== undefined) {
        const params = new URLSearchParams(search);
        nextPage = Number(params.get('page'));
        nextPerPage = Number(params.get('per_page'));
      } else {
        nextPage = getURIpage();
        nextPerPage = getURIperPage();
      }
    }
    setPerPage(current => nextPerPage || current || settingsPerPage);
    setPage(current => nextPage || current);
  }, [search, propsPage, propsPerPage, settingsPerPage, updateParamsByUrl]);

  const paginationTitles = {
    items: __('items'),
    page: '', // doesn't work well with translations as it adds 's' for plural, see: https://github.com/patternfly/patternfly-react/issues/6707
    itemsPerPage: __('Items per page'),
    perPageSuffix: __('per page'),
    toFirstPage: __('Go to first page'),
    toPreviousPage: __('Go to previous page'),
    toLastPage: __('Go to last page'),
    toNextPage: __('Go to next page'),
    optionsToggle: __('Items per page'),
    currPage: __('Current page'),
    paginationTitle: __('Pagination'),
  };

  const getPerPageOptions = () => {
    const options = new Set([5, 10, 15, 25, 50]);
    options.add(settingsPerPage);
    options.add(perPage);
    return [...options]
      .sort((a, b) => a - b)
      .map(value => ({ title: value.toString(), value }));
  };

  const _onSetPage = (e, nextPage) => {
    setPage(nextPage);
    if (onSetPage) return onSetPage(nextPage);
    const nextParams = { page: nextPage, per_page: perPage };
    if (onChange) return onChange(nextParams);
    return updateSearch(nextParams);
  };

  const _onPerPageSelect = (e, nextPerPage) => {
    setPerPage(nextPerPage);
    if (onPerPageSelect) return onPerPageSelect(nextPerPage);
    const nextParams = { per_page: nextPerPage, page: 1 };
    if (onChange) return onChange(nextParams);
    return updateSearch(nextParams);
  };

  const updateSearch = params => {
    if (!updateParamsByUrl) return;
    if (history) {
      const uri = new URI();
      uri.setSearch(params);
      history.push({ search: uri.search() });
    } else {
      changeQuery(params);
    }
  };

  const cx = classNames('tfm-pagination', className, {
    'no-side-padding': noSidePadding,
  });
  return (
    <PF4Pagination
      titles={paginationTitles}
      isCompact={variant === PaginationVariant.top}
      {...props}
      page={page}
      variant={variant}
      perPage={perPage}
      onSetPage={_onSetPage}
      onPerPageSelect={_onPerPageSelect}
      perPageOptions={getPerPageOptions()}
      data-per-page={perPage}
      data-total={itemCount}
      itemCount={itemCount}
      className={cx}
    />
  );
};

Pagination.propTypes = {
  onSetPage: PropTypes.func,
  onPerPageSelect: PropTypes.func,
  itemCount: PropTypes.number,
  className: PropTypes.string,
  page: PropTypes.number,
  perPage: PropTypes.number,
  noSidePadding: PropTypes.bool,
  variant: PropTypes.string,
  onChange: PropTypes.func,
  updateParamsByUrl: PropTypes.bool,
};

Pagination.defaultProps = {
  onSetPage: null,
  onPerPageSelect: null,
  onChange: null,
  itemCount: 0,
  className: null,
  page: 1,
  perPage: null,
  noSidePadding: false,
  variant: PaginationVariant.bottom,
  updateParamsByUrl: true,
};

export default Pagination;
