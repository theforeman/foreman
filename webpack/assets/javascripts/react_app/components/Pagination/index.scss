.pf-v5-c-pagination.pf-m-bottom.tfm-pagination {
  &.no-side-padding {
    padding: 16px 0;
  }
}
