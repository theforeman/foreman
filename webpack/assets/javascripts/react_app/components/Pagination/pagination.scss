#pagination {
  margin-top: -6px;

  .form-group {
    align-items: center;
  }

  span {
    margin-left: 5px;
  }
}
