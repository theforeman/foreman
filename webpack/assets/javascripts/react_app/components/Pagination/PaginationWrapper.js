import React from 'react';
import PropTypes from 'prop-types';
import { Paginator } from 'patternfly-react';
import Pagination from './Pagination';

const PaginationWrapper = props => {
  const {
    onPageSet,
    onPerPageSelect,
    onChange,
    pagination,
    dropdownButtonId,
    itemCount,
    viewType,
    ...otherProps
  } = props;

  const onPageSetUpdate = page => {
    update({ page });
    onPageSet(page);
  };

  const onPerPageSelectUpdate = perPage => {
    update({ perPage, page: 1 });
    onPerPageSelect(perPage);
  };

  const update = changes => {
    const newPagination = { ...pagination, ...changes };

    onChange({
      page: newPagination.page,
      perPage: newPagination.perPage,
    });
  };

  const initPagination = paginationFromProps => {
    const defaultPagination = {
      page: 1,
      perPage: 20,
      perPageOptions: [5, 10, 15, 20, 25, 50],
    };
    return { ...defaultPagination, ...(paginationFromProps || {}) };
  };

  return (
    <Pagination
      data={{ itemCount, viewType }}
      onPageSet={onPageSetUpdate}
      onPerPageSelect={onPerPageSelectUpdate}
      dropdownButtonId={dropdownButtonId}
      pagination={initPagination(pagination)}
      {...otherProps}
    />
  );
};

PaginationWrapper.defaultProps = {
  onChange: () => {},
  viewType: 'list',
  ...Paginator.defaultProps,
};

PaginationWrapper.propTypes = {
  ...Paginator.propTypes,
  /** page and per-page selection callback */
  onChange: PropTypes.func,
  /** view type */
  viewType: PropTypes.string,
  /** pagination */
  pagination: PropTypes.shape({
    /** the current page */
    page: PropTypes.number,
    /** the current per page setting */
    perPage: PropTypes.number,
    /** per page options */
    perPageOptions: PropTypes.arrayOf(PropTypes.number),
  }),
};

export default PaginationWrapper;
