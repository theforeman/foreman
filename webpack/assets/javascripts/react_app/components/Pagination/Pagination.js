import React from 'react';
import Proptypes from 'prop-types';
import { isEmpty } from 'lodash';
import { Paginator } from 'patternfly-react';
import {
  getURI,
  getURIpage,
  getURIperPage,
  changeQuery,
  translatePagination,
} from './PaginationHelper';
import './pagination.scss';

const Pagination = (props) => {
  const {
    data,
    pagination,
    onPageSet,
    onPerPageSelect,
    dropdownButtonId,
    ...otherProps
  } = props;

  const urlPage = getURIpage();
  const urlPerPage = getURIperPage();
  const className = isEmpty(data.classNames)
    ? 'col-md-12'
    : `col-md-12 ${data.classNames.pagination_classes}`;

  return (
    <Paginator
      pagination={
        isEmpty(pagination)
          ? {
              page: urlPage || 1,
              perPage: urlPerPage || data.perPage,
              perPageOptions: data.perPageOptions,
            }
          : pagination
      }
      viewType={data.viewType}
      itemCount={data.itemCount}
      onPageSet={onPageSet}
      onPerPageSelect={onPerPageSelect}
      messages={translatePagination(Paginator.defaultProps.messages)}
      className={className}
      dropdownButtonId={dropdownButtonId}
      { ...otherProps }
    />
  );
};

Pagination.propTypes = {
  data: Proptypes.shape({
    viewType: Proptypes.string,
    perPageOptions: Proptypes.arrayOf(Proptypes.number),
    itemCount: Proptypes.number,
    perPage: Proptypes.number,
  }).isRequired,
  onPageSet: Proptypes.func,
  onPerPageSelect: Proptypes.func,
  dropdownButtonId: Proptypes.string,
  pagination: Proptypes.shape({
    page: Proptypes.number,
    perPage: Proptypes.number,
    perPageOptions: Proptypes.arrayOf(Proptypes.number),
  }),
};

Pagination.defaultProps = {
  onPageSet: page => changeQuery(getURI(), { page }),
  onPerPageSelect: perPage => changeQuery(getURI(), { per_page: perPage }),
  dropdownButtonId: 'pagination-row-dropdown',
  pagination: null,
};

export default Pagination;
