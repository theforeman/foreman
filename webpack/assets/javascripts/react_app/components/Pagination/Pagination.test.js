import { testComponentSnapshotsWithFixtures } from '../../common/testHelpers';
import Pagination from '../Pagination/Pagination';
import PaginationWrapper from '../Pagination/PaginationWrapper';

import { paginationMock } from './Pagination.fixtures';

const fixtures = {
  'renders layout': paginationMock,
};

const getBaseProps = () => ({
  pagination: {
    page: 2,
    perPage: 5,
    perPageOptions: [5, 10, 25],
  },
  itemCount: 52,
  viewType: 'list',
});

describe('Pagination', () => {
  describe('rendering from erb', () =>
    testComponentSnapshotsWithFixtures(Pagination, fixtures));

  describe('rendering from js', () =>
    testComponentSnapshotsWithFixtures(PaginationWrapper, {
      'renders correctly': getBaseProps(),
    }));
});
