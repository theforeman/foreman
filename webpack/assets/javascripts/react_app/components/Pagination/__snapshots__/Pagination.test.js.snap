// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination rendering from erb renders layout 1`] = `
<Paginator
  className="col-md-12"
  dropdownButtonId="pagination-row-dropdown"
  itemCount={21}
  messages={
    Object {
      "currentPage": "Current Page",
      "firstPage": "First Page",
      "lastPage": "Last Page",
      "nextPage": "Next Page",
      "of": "of",
      "perPage": "per page",
      "previousPage": "Previous Page",
    }
  }
  onPageSet={[Function]}
  onPerPageSelect={[Function]}
  pagination={
    Object {
      "page": 1,
      "perPage": 20,
      "perPageOptions": Array [
        5,
        10,
        15,
        20,
      ],
    }
  }
  viewType="table"
/>
`;

exports[`Pagination rendering from js renders correctly 1`] = `
<Pagination
  className=""
  data={
    Object {
      "itemCount": 52,
      "viewType": "list",
    }
  }
  dropdownButtonId="pagination-row-dropdown"
  messages={
    Object {
      "currentPage": "Current Page",
      "firstPage": "First Page",
      "lastPage": "Last Page",
      "nextPage": "Next Page",
      "of": "of",
      "perPage": "per page",
      "previousPage": "Previous Page",
    }
  }
  onPageSet={[Function]}
  onPerPageSelect={[Function]}
  pagination={
    Object {
      "page": 2,
      "perPage": 5,
      "perPageOptions": Array [
        5,
        10,
        25,
      ],
    }
  }
/>
`;
